### PR TITLE
Integrate MMS inferenceservice controller, configmap controller, model agent

### PR DESCRIPTION
**Pull request by yuzliu**
**What this PR does / why we need it**:
This change is for card: https://github.com/kubeflow/kfserving/projects/3#card-44938474

1) When an InferenceService created, if the storageUri is empty the InferenceService controller will create a configmap for models 
2) Added an agentInjector so the mutator will inject model agent to predictor pod if the storageUri is not set
3) Mount modelConfig into predictor pod and use agent container if the storageUri is not set
4) Do not mount storageInitializer sidecar when the InferenceService is a MMS
5) Delete TrainedModels if their parent InferenceService is deleted.
6) Fixed a bug in agent so it does not unload models that didn't change
7) Add a finalizer for TrainedModel, so before a TrainedModel is removed it will remove itself from the modelConfig
8) Add a finalizer for InferenceService, so when an InferenceService is delete the controller will delete all the TrainedModels that belong to a parent InferenceService.
9) InferenceService controller will only reconcile model config and inject agent side for for MMS enabled model server. Currently only SKLearn and XGBoost are enabled for MMS.

Testing process

Unset all proxy
```
  unset http_proxy
  unset https_proxy
  unset HTTPS_PROXY
  unset HTTP_PROXY
```

Configure KO
```
export KO_DOCKER_REPO=docker.io/<docker login>
```
Modify .ko.yaml to change the base image for agent so the base image has debugging tools
```
baseImageOverrides:
    github.com/kubeflow/kfserving/cmd/agent: docker.io/library/ubuntu:latest
```

Create a testing kind cluster for Kubernetes 1.16
```
kind create cluster --image=kindest/node:v1.16.9@sha256:7175872357bc85847ec4b1aba46ed1d12fa054c83ac7a8a11f5c268957fd5765
```

Publish the SKLearn model server
```
cd /Users/yliu989/go-workspace/src/github.com/kubeflow/kfserving/python
docker build -t yuzhui/sklearnserver:mms -f sklearn.Dockerfile .
docker push yuzhui/sklearnserver:mms
```

Update image_patch_dev.sh so it patches model server
```
cat > config/overlays/${OVERLAY}/configmap/inferenceservice_patch.yaml << EOF
apiVersion: v1
kind: ConfigMap
metadata:
 name: inferenceservice-config
 namespace: kfserving-system
data:
 logger: |-
   {
       "image" : "${LOGGER_IMG}",
       "memoryRequest": "100Mi",
       "memoryLimit": "1Gi",
       "cpuRequest": "100m",
       "cpuLimit": "1"
   }
 batcher: |-
   {
       "image" : "${BATCHER_IMG}",
       "memoryRequest": "100Mi",
       "memoryLimit": "1Gi",
       "cpuRequest": "100m",
       "cpuLimit": "1"
   }
 agent: |-
   {
       "image" : "${AGENT_IMG}",
       "memoryRequest": "100Mi",
       "memoryLimit": "1Gi",
       "cpuRequest": "100m",
       "cpuLimit": "1"
   }
 predictors: |-
   {
     "sklearn": {
         "v1": {
           "image": "yuzhui/sklearnserver",
           "defaultImageVersion": "fixstorage3",
           "supportedFrameworks": [
             "sklearn"
           ],
           "multiModelServer": "false"
         },
         "v2": {
           "image": "docker.io/seldonio/mlserver",
           "defaultImageVersion": "0.1.2",
           "supportedFrameworks": [
             "sklearn"
           ],
           "multiModelServer": "false"
         }
       },
   }
```

Deploy latest branch
```
make deploy-dev
```

Deploy a mock S3
```
cd ./docs/samples/kafka
kubectl apply -f minio.yaml
kubectl port-forward $(kubectl get pod --selector="app=minio" --output jsonpath='{.items[0].metadata.name}') 9000:9000
#Upload models to the mock S3
```

Deploy S3 secret
```
apiVersion: v1
kind: Secret
metadata:
  name: mysecret
  annotations:
     serving.kubeflow.org/s3-endpoint: "minio-service.default:9000" # replace with your s3 endpoint
     serving.kubeflow.org/s3-usehttps: "0" # by default 1, for testing with minio you need to set to 0
     serving.kubeflow.org/s3-region: "us-west-2"
     serving.kubeflow.org/s3-usevirtualbucket: "false"
type: Opaque
data:
  AWS_ACCESS_KEY_ID: bWluaW8=
  AWS_SECRET_ACCESS_KEY: bWluaW8xMjM=
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: sa
secrets:
  - name: mysecret

```
Create an InferenceService without storageUri and S3 annotations
```
apiVersion: "serving.kubeflow.org/v1beta1"
kind: "InferenceService"
metadata:
  name: "sklearn-iris"
spec:
  predictor:
    serviceAccountName: "sa"
    sklearn:
      runtimeVersion: "mms"
      name: "sklearn-iris-predictor"
      resources:
        limits:
          cpu: "0.2"
          memory: 200Mi
        requests:
          cpu: "0.2"
          memory: 200Mi
```

Verified that configmap is created for every InferenceService
```
k get cm modelconfig-sklearn-iris-0  -oyaml
apiVersion: v1
kind: ConfigMap
metadata:
  creationTimestamp: "2020-09-05T06:32:27Z"
  name: modelconfig-sklearn-iris-0
  namespace: default
  ownerReferences:
  - apiVersion: serving.kubeflow.org/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: InferenceService
    name: sklearn-iris
    uid: 3b1f17ea-0afa-49c7-a1c8-ad3a08821c2e
  resourceVersion: "747575"
  selfLink: /api/v1/namespaces/default/configmaps/modelconfig-sklearn-iris-0
  uid: 24de7a99-36c1-4a59-bd32-697393f67909
```
Verified that model agent container is mounted in predictor pod and the modelconfig volume is mounted
```
  - image: docker.io/yuzhui/agent:latest
    imagePullPolicy: Always
    name: agent
    resources:
      limits:
        cpu: "1"
        memory: 1Gi
      requests:
        cpu: "1"
        memory: 1Gi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /mnt/configs
      name: model-config
      readOnly: true
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: default-token-vxz2p
      readOnly: true
```

Create TrainedModel
```
apiVersion: "serving.kubeflow.org/v1beta1"
kind: "TrainedModel"
metadata:
  name: "model1"
spec:
  inferenceService: "sklearn-iris"
  model:
    storageUri: "s3://kfserving-test/model1"
    framework: "sklearn"
    memory: "1G"
---
apiVersion: "serving.kubeflow.org/v1beta1"
kind: "TrainedModel"
metadata:
  name: "model2"
spec:
  inferenceService: "sklearn-iris"
  model:
    storageUri: "s3://kfserving-test/model2"
    framework: "sklearn"
    memory: "1G"
```

Verify configMap is updated
```
apiVersion: v1
data:
  models.json: '[{"modelName":"model1","modelSpec":{"storageUri":"s3://kfserving-test/model1","framework":"sklearn","memory":"1G"}}]'
kind: ConfigMap
metadata:
  creationTimestamp: "2020-10-22T02:48:32Z"
  name: modelconfig-sklearn-iris-0
  namespace: default
  ownerReferences:
  - apiVersion: serving.kubeflow.org/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: InferenceService
    name: sklearn-iris
    uid: 6f4a0ee5-5705-420f-99fc-99e26b26f441
  resourceVersion: "157489"
  selfLink: /api/v1/namespaces/default/configmaps/modelconfig-sklearn-iris-0
  uid: 2f6be618-5abd-4499-a1de-05c1be9932d9
```

Verify you can curl each model to run predictions
```
INGRESS_GATEWAY_SERVICE=$(kubectl get svc --namespace istio-system --selector="app=istio-ingressgateway" --output jsonpath='{.items[0].metadata.name}')


kubectl port-forward --namespace istio-system svc/${INGRESS_GATEWAY_SERVICE} 8080:80
export INGRESS_HOST=localhost
export INGRESS_PORT=8080

SERVICE_HOSTNAME=$(kubectl get inferenceservice sklearn-iris -n default -o jsonpath='{.status.url}' | cut -d "/" -f 3)

curl -v -H "Host: ${SERVICE_HOSTNAME}" http://${INGRESS_HOST}:${INGRESS_PORT}/v1/models/model1:predict -d @./docs/samples/sklearn/iris-input.json

curl -v -H "Host: ${SERVICE_HOSTNAME}" http://${INGRESS_HOST}:${INGRESS_PORT}/v1/models/model1:predict -d @./docs/samples/sklearn/iris-input.json

```

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.
No image changes

**Release note**:

```release-note

```

### CODE REVIEW

File: pkg/controller/v1beta1/inferenceservice/components/explainer.go [hunk 1]
```diff
@@ -85,7 +85,7 @@ func (p *Explainer) Reconcile(isvc *v1beta1.InferenceService) error {
 
 	podSpec := v1.PodSpec(isvc.Spec.Explainer.PodSpec)
 	r := knative.NewKsvcReconciler(p.client, p.scheme, objectMeta, &isvc.Spec.Explainer.ComponentExtensionSpec,
-		&podSpec, isvc.Status.Components[v1beta1.ExplainerComponent])
+		&podSpec, isvc.Status.Components[v1beta1.ExplainerComponent], constants.Explainer, isvc)
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/components/explainer.go:88]: I think we do not need these two extra arguments, the knative service reconciler logic is not component specific.

File: pkg/controller/v1beta1/inferenceservice/controller.go [hunk 2]
```diff
@@ -105,6 +107,12 @@ func (r *InferenceServiceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, e
 		return reconcile.Result{}, errors.Wrapf(err, "fails to reconcile ingress")
 	}
 
+	//Reconcile modelConfig
+	configMapReconciler := modelconfig.NewModelConfigReconciler(r.Client, r.Scheme)
+	if err := configMapReconciler.Reconcile(isvc, req); err != nil {
+		return reconcile.Result{Requeue: true, RequeueAfter: 10 * time.Second}, err
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/controller.go:113]: controller runtime should be able to handle the retry if err is returned

File: pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go [hunk 3]
```diff
@@ -76,6 +86,9 @@ func createKnativeService(componentMeta metav1.ObjectMeta,
 	if _, ok := annotations[autoscaling.ClassAnnotationKey]; !ok {
 		annotations[autoscaling.ClassAnnotationKey] = autoscaling.KPA
 	}
+	// Add agent annotations so mutator will mount model agent to multi-model InferenceService's predictor
+	addAgentAnnotations(component, isvc, annotations)
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go:90]: this logic should be in `predictor.go`

File: pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go [hunk 4]
```diff
@@ -134,6 +147,26 @@ func createKnativeService(componentMeta metav1.ObjectMeta,
 	return service
 }
 
+func addAgentAnnotations(component constants.InferenceServiceComponent, isvc *v1beta1.InferenceService, annotations map[string]string) bool {
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go:150]: same here this function needs to be moved to `predictor.go`

File: pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go [hunk 5]
```diff
@@ -134,6 +147,26 @@ func createKnativeService(componentMeta metav1.ObjectMeta,
...
+	if component == constants.Predictor && isvc.Spec.Predictor.GetImplementation().GetStorageUri() == nil {
+		annotations[constants.AgentInternalAnnotationKey] = "true"
+		shardStrategy := memory.MemoryStrategy{}
+		for _, id := range shardStrategy.GetShard(isvc) {
+			modelConfig, err := modelconfig.CreateEmptyModelConfig(isvc, id)
+			if err == nil {
+				annotations[constants.AgentModelConfigAnnotationKey] = modelConfig.Name
+				//TODO figure out the proper way to inject s3 endpoint from s3 identity
+				annotations[constants.AgentS3endpointAnnotationKey] = "gs://kfserving-samples"
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go:159]: Why is this needed? The s3 endpoint should be injected in the container in the agent sidecar.

File: pkg/webhook/admission/pod/agent_injector.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+	pod.Spec.Volumes = appendVolume(pod.Spec.Volumes, additionalVolume)
+	var mountedContainers []v1.Container
+	for _, container := range pod.Spec.Containers {
+		if container.Name == containerName {
+			if container.VolumeMounts == nil {
+				container.VolumeMounts = []v1.VolumeMount{}
+			}
+			container.VolumeMounts = append(container.VolumeMounts, v1.VolumeMount{
+				Name:      additionalVolume.Name,
+				ReadOnly:  true,
```
Comment (yuzisun) [inline pkg/webhook/admission/pod/agent_injector.go:168]: I think model directory is not read only?

File: test/workflows/components/workflows.libsonnet [hunk 7]
```diff
@@ -249,6 +249,10 @@
                     name: "build-batcher",
                     template: "build-batcher",
                   },
+                  {
+                    name: "build-agent",
+                    template: "build-batcher",
```
Comment (yuzisun) [inline test/workflows/components/workflows.libsonnet:254]: build-agent?

File: python/kfserving/kfserving/constants/constants.py [hunk 8]
```diff
@@ -18,7 +18,7 @@
 KFSERVING_GROUP = 'serving.kubeflow.org'
 KFSERVING_KIND = 'InferenceService'
 KFSERVING_PLURAL = 'inferenceservices'
-KFSERVING_VERSION = os.environ.get('KFSERVING_VERSION', 'v1alpha2')
+KFSERVING_VERSION = os.environ.get('KFSERVING_VERSION', 'v1beta1')
```
Comment (yuzisun) [inline python/kfserving/kfserving/constants/constants.py:21]: Let's still keep the default to be v1alpha2 as all tests are not converted to v1beta1 yet.

File: pkg/webhook/admission/pod/agent_injector.go [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+	//Ensure that we set proper values for CPU/Memory Limit/Request
+	resourceDefaults := []string{agentConfig.MemoryRequest,
+		agentConfig.MemoryLimit,
+		agentConfig.CpuRequest,
+		agentConfig.CpuLimit}
+	for _, key := range resourceDefaults {
+		_, err := resource.ParseQuantity(key)
+		if err != nil {
+			return agentConfig, fmt.Errorf("Failed to parse resource configuration for %q: %q",
+				BatcherConfigMapKeyName, err.Error())
```
Comment (yuzisun) [inline pkg/webhook/admission/pod/agent_injector.go:59]: AgentConfigMapKeyName?

File: pkg/webhook/admission/pod/agent_injector_test.go [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+
+	"github.com/kubeflow/kfserving/pkg/constants"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	agentConfig = &AgentConfig{
+		Image:         "gcr.io/kfserving/agent:latest",
+		CpuRequest:    BatcherDefaultCPURequest,
```
Comment (yuzisun) [inline pkg/webhook/admission/pod/agent_injector_test.go:32]: AgentDefaultCPURequest?

File: pkg/controller/v1beta1/inferenceservice/components/explainer.go (hunk 1 above)
Comment (yuzliu) [reply pkg/controller/v1beta1/inferenceservice/components/explainer.go:88]: In the ksvc_reconciler I need to check the component type and only inject agent sidecar to predictor:
```
func addAgentAnnotations(component constants.InferenceServiceComponent, isvc *v1beta1.InferenceService, annotations map[string]string) bool {
	if component == constants.Predictor && isvc.Spec.Predictor.GetImplementation().GetStorageUri() == nil {
```

File: pkg/controller/v1beta1/inferenceservice/controller.go (hunk 2 above)
Comment (yuzliu) [reply pkg/controller/v1beta1/inferenceservice/controller.go:113]: Changed to ```
		return reconcile.Result{}, err

```

File: pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go (hunk 5 above)
Comment (yuzliu) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go:159]: Oops forgot to delete this line, will remove.

File: pkg/webhook/admission/pod/agent_injector.go (hunk 6 above)
Comment (yuzliu) [reply pkg/webhook/admission/pod/agent_injector.go:168]: Agree, should set it to false.

File: test/workflows/components/workflows.libsonnet (hunk 7 above)
Comment (yuzliu) [reply test/workflows/components/workflows.libsonnet:254]: updated

File: python/kfserving/kfserving/constants/constants.py (hunk 8 above)
Comment (yuzliu) [reply python/kfserving/kfserving/constants/constants.py]: Will change.

File: pkg/webhook/admission/pod/agent_injector_test.go (hunk 10 above)
Comment (yuzliu) [reply pkg/webhook/admission/pod/agent_injector_test.go:32]: Will change.

File: pkg/webhook/admission/pod/agent_injector.go (hunk 9 above)
Comment (yuzliu) [reply pkg/webhook/admission/pod/agent_injector.go:59]: Will change.

File: hack/image_patch_dev.sh [hunk 11]
```diff
@@ -49,4 +52,32 @@ data:
...
+        "image" : "${AGENT_IMG}",
+        "memoryRequest": "100Mi",
+        "memoryLimit": "1Gi",
+        "cpuRequest": "100m",
+        "cpuLimit": "1"
+    }
+  predictors: |-
+   {
+       "sklearn": {
+           "image": "yuzhui/sklearnserver",
```
Comment (ifilonenko) [inline hack/image_patch_dev.sh:66]: this should be removed

File: pkg/controller/v1beta1/inferenceservice/controller.go [hunk 12]
```diff
@@ -47,7 +47,7 @@ import (
 // +kubebuilder:rbac:groups=networking.istio.io,resources=virtualservices/finalizers,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=core,resources=serviceaccounts,verbs=get;list;watch
-// +kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;list;watch;create
```
Comment (ifilonenko) [inline pkg/controller/v1beta1/inferenceservice/controller.go:50]: update?

File: python/kfserving/kfserving/storage.py [hunk 13]
```diff
@@ -63,6 +64,10 @@ def download(uri: str, out_dir: str = None) -> str:
             return Storage._download_local(uri, out_dir)
         elif re.search(_URI_RE, uri):
             return Storage._download_from_uri(uri, out_dir)
+        elif uri.startswith(MODEL_MOUNT_DIRS):
```
Comment (yuzisun) [inline python/kfserving/kfserving/storage.py:67]: for MMS, we do not even run the storage initializer right?

Comment (yuzliu) [reply python/kfserving/kfserving/storage.py:67]: The sklearn server also uses storage.py to download models. This change it to make it compatible with MMS. Otherwise, the sklearn server container will crash.

Comment (yuzisun) [reply python/kfserving/kfserving/storage.py:67]: hmm I thought it should hit line 52.

File: test/workflows/components/workflows.libsonnet [hunk 14]
```diff
@@ -249,6 +249,10 @@
                     name: "build-batcher",
                     template: "build-batcher",
                   },
+                  {
+                    name: "build-agent",
+                    template: "build-agent",
```
Comment (yuzisun) [inline test/workflows/components/workflows.libsonnet:254]: I think it is missing the build template for `build-agent`, that's why the argo workflow is not triggered

Comment (yuzliu) [reply test/workflows/components/workflows.libsonnet:254]: Thanks, will fix!

File: python/kfserving/kfserving/storage.py (hunk 13 above)
Comment (yuzliu) [reply python/kfserving/kfserving/storage.py:67]: It will only hit line 52 if the storageUri starts with ```file://```, I guess that's for PVC?

File: hack/image_patch_dev.sh (hunk 11 above)
Comment (yuzliu) [reply hack/image_patch_dev.sh:66]: Thanks, will do.

File: pkg/controller/v1beta1/inferenceservice/components/predictor.go [hunk 15]
```diff
@@ -168,3 +172,22 @@ func addBatcherContainerPort(container *v1.Container) {
 		}
 	}
 }
+
+func addAgentAnnotations(isvc *v1beta1.InferenceService, annotations map[string]string) bool {
+	if isvc.Spec.Predictor.GetImplementation().GetStorageUri() == nil {
+		annotations[constants.AgentShouldInjectAnnotationKey] = "true"
+		shardStrategy := memory.MemoryStrategy{}
+		for _, id := range shardStrategy.GetShard(isvc) {
+			modelConfig, err := modelconfig.CreateEmptyModelConfig(isvc, id)
```
[the remark this answers is not in the document]
Comment (yuzliu) [reply pkg/controller/v1beta1/inferenceservice/components/predictor.go:181]: Right, I only need to construct the configMap name here. Will change.

File: pkg/controller/v1beta1/inferenceservice/controller.go (hunk 12 above)
Comment (yuzliu) [reply pkg/controller/v1beta1/inferenceservice/controller.go:51]: InferenceService controller won't update the configmap, only the trainedmodel controller will update.

File: hack/image_patch_dev.sh (hunk 11 above)
Comment (yuzliu) [reply hack/image_patch_dev.sh]: Removed.

File: pkg/agent/downloader.go [hunk 16]
```diff
@@ -47,6 +47,7 @@ func (d *Downloader) DownloadModel(modelName string, modelSpec *v1beta1.ModelSpe
 		log.Info("Downloading to model dir", "modelUri", modelUri, "modelDir", d.ModelDir)
 		// Download if the event there is a success file and the event is one which we wish to Download
 		if !storage.FileExists(successFile) {
+			log.Info("Success file does not exist", "successfile", successFile)
```
Comment (ifilonenko) [inline pkg/agent/downloader.go:50]: Is this necessary to log?

File: cmd/agent/main.go [hunk 17]
```diff
@@ -10,25 +10,39 @@ import (
...
 var (
+	log       = logf.Log.WithName("modelAgent")
 	configDir = flag.String("config-dir", "/mnt/configs", "directory for model config files")
 	modelDir  = flag.String("model-dir", "/mnt/models", "directory for model files")
 )
 
 func main() {
 	flag.Parse()
+	logf.SetLogger(logf.ZapLogger(false))
+	//filepath.
```
Comment (yuzisun) [inline cmd/agent/main.go:26]: remove this line?

File: cmd/manager/main.go [hunk 18]
```diff
@@ -71,7 +72,10 @@ func main() {
 
 	// Create a new Cmd to provide shared dependencies and start components
 	log.Info("Setting up manager")
-	mgr, err := manager.New(cfg, manager.Options{MetricsBindAddress: metricsAddr, Port: 9443})
+	// Setting trainedmodel's resync period to 60s
+	var syncPeriod = 60 * time.Second
```
Comment (yuzisun) [inline cmd/manager/main.go:76]: we do not need this resync any more if we delete the trained model after inference service is gone right?

File: pkg/agent/puller.go [hunk 19]
```diff
@@ -79,6 +81,7 @@ type ModelChannel struct {
 }
 
 func (p *Puller) enqueueModelOp(modelOp *ModelOp) {
+	log.Info("enqueue", "modelop", modelOp)
```
Comment (yuzisun) [inline pkg/agent/puller.go:84]: debug level?

File: pkg/agent/puller.go [hunk 20]
```diff
@@ -59,6 +59,8 @@ func StartPuller(downloader Downloader, commands <-chan ModelOp) {
 
 func (p *Puller) processCommands(commands <-chan ModelOp) {
 	// channelMap accessed only by this goroutine
+	log.Info("Start to process model commands")
```
Comment (yuzisun) [inline pkg/agent/puller.go:62]: probably can remove this logging now

File: pkg/controller/v1beta1/inferenceservice/controller.go [hunk 21]
```diff
@@ -105,6 +107,12 @@ func (r *InferenceServiceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, e
 		return reconcile.Result{}, errors.Wrapf(err, "fails to reconcile ingress")
 	}
 
+	//Reconcile modelConfig
+	configMapReconciler := modelconfig.NewModelConfigReconciler(r.Client, r.Scheme)
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/controller.go:111]: Can we move this to predictor as it is only relevant for model server

File: pkg/controller/v1beta1/inferenceservice/components/explainer.go (hunk 1 above)
Comment (yuzliu) [reply pkg/controller/v1beta1/inferenceservice/components/explainer.go]: I moved the logic that adding agent annotation to the predictor so I no longer need to pass extra variables.

File: pkg/credentials/s3/s3_secret.go [hunk 22]
```diff
@@ -127,10 +131,23 @@ func BuildSecretEnvs(secret *v1.Secret, s3Config *S3Config) []v1.EnvVar {
...
+		})
+	}
+
+	if useVirtualBucket, ok := secret.Annotations[InferenceServiceS3UseVirtualBucketAnnotation]; ok {
+		envs = append(envs, v1.EnvVar{
+			Name:  S3UseVirtualBucket,
+			Value: useVirtualBucket,
+		})
+	} else {
+		envs = append(envs, v1.EnvVar{
```
[the remark this answers is not in the document]
Comment (yuzliu) [reply pkg/credentials/s3/s3_secret.go:147]: Ok I removed the default logic

File: pkg/agent/downloader.go (hunk 16 above)
Comment (yuzliu) [reply pkg/agent/downloader.go:50]: I switched to use log level 9

File: pkg/agent/puller.go (hunk 19 above)
Comment (yuzliu) [reply pkg/agent/puller.go:84]: Used log level 10

File: pkg/agent/puller.go (hunk 20 above)
Comment (yuzliu) [reply pkg/agent/puller.go:62]: Removed

File: pkg/controller/v1beta1/inferenceservice/controller.go (hunk 21 above)
Comment (yuzliu) [reply pkg/controller/v1beta1/inferenceservice/controller.go:111]: There won't be any InferenceService without a predictor right?

File: test/workflows/components/workflows.libsonnet [hunk 23]
```diff
@@ -372,6 +376,12 @@
               "--context=dir://" + srcDir,
               "--destination=" + "527798164940.dkr.ecr.us-west-2.amazonaws.com/kfserving/batcher:$(PULL_BASE_SHA)",
             ]),  // build-batcher
+            $.parts(namespace, name, overrides).e2e(prow_env, bucket).buildTemplate("build-agent", kanikoExecutorImage, [
+              "/kaniko/executor",
+              "--dockerfile=" + srcDir + "/agent.Dockerfile",
+              "--context=dir://" + srcDir,
+              "--destination=" + "527798164940.dkr.ecr.us-west-2.amazonaws.com/kfserving/agent:$(PULL_BASE_SHA)",
```
Comment (yuzisun) [inline test/workflows/components/workflows.libsonnet:383]: @PatrickXYS Can you help add the above image to ECR registry? thanks

File: pkg/controller/v1beta1/trainedmodel/controller.go [hunk 24]
```diff
@@ -64,22 +63,66 @@ func (r *TrainedModelReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
...
+		if errors.IsNotFound(err) {
+			log.Info("Parent InferenceService does not exists, deleting TrainedModel", "TrainedModel", tm.Name, "InferenceService", isvc.Name)
+			r.Delete(context.TODO(), tm)
+			return reconcile.Result{}, nil
+		}
 		return reconcile.Result{}, err
 	}
+	// Set TrainedModel's ownerReference to its parent InferenceService so when the parent InferenceService is removed
+	// the TraineModel object will be removed
+	log.Info("add owner reference")
```
Comment (yuzisun) [inline pkg/controller/v1beta1/trainedmodel/controller.go:79]: this can be removed?

File: pkg/controller/v1beta1/trainedmodel/controller.go [hunk 25]
```diff
@@ -64,22 +63,66 @@ func (r *TrainedModelReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
...
+	if err := controllerutil.SetControllerReference(isvc, tm, r.Scheme); err != nil {
```
Comment (yuzisun) [inline pkg/controller/v1beta1/trainedmodel/controller.go:80]: Did you check if the owner reference is set correctly, I believe you will need to do an update here for the isvc.

File: python/kfserving/kfserving/storage.py (hunk 13 above)
Comment (yuzisun) [reply python/kfserving/kfserving/storage.py:67]: still wondering why `os.path.exists(uri)` did not evaluate to be true for MMS case.

File: pkg/controller/v1beta1/inferenceservice/controller.go (hunk 21 above)
Comment (yuzisun) [reply pkg/controller/v1beta1/inferenceservice/controller.go:111]: let's add a check only to create the config if storageUri is nil, so user are not confused creating an empty config

Comment (yuzliu) [reply pkg/controller/v1beta1/inferenceservice/controller.go:111]: Done

File: pkg/controller/v1beta1/trainedmodel/controller.go (hunk 24 above)
Comment (yuzliu) [reply pkg/controller/v1beta1/trainedmodel/controller.go:79]: Removed.

File: pkg/controller/v1beta1/trainedmodel/controller.go (hunk 25 above)
Comment (yuzliu) [reply pkg/controller/v1beta1/trainedmodel/controller.go:80]: The ownerreference is added but when I delete InferenceService the trainedmodel was not removed. Let me try adding an update.

Comment (yuzliu) [reply pkg/controller/v1beta1/trainedmodel/controller.go:80]: I didn't need to send an update for isvc when I set ownerReference for model configmap and the configmap is garbage collected correctly:
```
if err := controllerutil.SetControllerReference(isvc, newModelConfig, c.scheme); err != nil {
	return err
}
```

File: pkg/controller/v1beta1/inferenceservice/components/predictor.go [hunk 26]
```diff
@@ -168,3 +171,18 @@ func addBatcherContainerPort(container *v1.Container) {
 		}
 	}
 }
+
+func addAgentAnnotations(isvc *v1beta1.InferenceService, annotations map[string]string) bool {
+	if isvc.Spec.Predictor.GetImplementation().GetStorageUri() == nil {
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/components/predictor.go:176]: For custom predictor this can be true, we need additional check to exclude custom predictor case(`len(isvc.Spec.Predictor.PodSpec.Containers) != 0`).

**Comment by yuzisun** [comment]
This is exciting! Great job @yuzliu !

/lgtm
/approve
